adding RFC process-handle-for-async (#2823)
rust-lang/rfcs: text/0000-process-handle-for-async.md

- Feature Name: `process_handle_for_async`
- Start Date: 2019-11-28
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Introduce process descriptor handles for each process in `'std::process'` for async polling of process-termination.

# Motivation
[motivation]: #motivation?

Currently `std::process` is managing processes (on Posix/Linux systems) either in synchronous way using the wait-function-family, or asynchronously via signal-handlers (SIGCHLD). 

The problem is that such signals may be delivered to any thread running within the parent-process. So, dealing with the termination of child-processes in async manner will require a shared collection containing the process-status of all child-processes.

This RFC proposes a portable alternative to signal-handling to deal with the termination of child-processes in async manner, and introducing strict ownership for child-processe using a process-handle (aka process descriptor)

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> Depending on the underlying operating system, sub-processes may be associated with a process-id (pid) and most oeprating systems provide  so called process-handles (process descriptors). Such handles may be used to perform async polling or sending signals to the corresponding process. Such process handles may be used to establish strict opwnership for child-processes in Rust.

Reviewer: ```suggestion
Depending on the underlying operating system, sub-processes may be associated with a process-id (pid) and most operating systems provide  so called process-handles (process descriptors). Such handles may be used to perform async polling or sending signals to the corresponding process. Such process handles may be used to establish strict ownership for child-processes in Rust.
```

> The process-handle feature shall permita portable API for various operating systems. The following listing shows the process-handle features being available for different operating systems, and might be the base for a generalized Rust-API in `std::process`.

Reviewer: ```suggestion
The process-handle feature shall permit a portable API for various operating systems. The following listing shows the process-handle features being available for different operating systems, and might be the base for a generalized Rust-API in `std::process`.
```

- As with Windows, the fucntion `CreateProcess` provides the HANDLE for the newly created child-process.
- As with OpenBSD Unix provides process handles 
- As with FreeBSD Unix the function `libc::pdfork(..)` provides a process descriptor for the newly forked child-process.
- As with Linux, the `pidfd`-familiy provides funtionality to convert a pid int a file-descriptor (merged into release Linux-5.4)
- Older Linux-releases or Posix-Systems may use the `forkfd`-concept to establsih a descriptor/handle for a child-process
 

> As for the latter case when the underlying OS does not provide a suitable naticve process-handle, the forkfd concept may be used. This is based on so called death-pipes: An anonymous pipe is established between parent-process (RX end) and the child-process (TX end). Here the parent-process may use RX end to poll for read-events (file descriptor) in async manner, for example using a future.

Reviewer: ```suggestion
As for the latter case when the underlying OS does not provide a suitable native process-handle, the forkfd concept may be used. This is based on so called death-pipes: An anonymous pipe is established between parent-process (RX end) and the child-process (TX end). Here the parent-process may use RX end to poll for read-events (file descriptor) in async manner, for example using a future.
```

Now, if the child-process terminates the TX end is being destroyed and the RX-end in the parent process will receive and EOF read-event.

> All of the listed process-handle flavors share the common concept to perform async polling for process-termination and might be the base for a generalized process-handle concpet in Rust `std::process`, leveraging a strong ownership-concpet for child-processes. The handling of signal SIGCHLD and sharing of child-process meta-data would no longer be required.

Reviewer: ```suggestion
All of the listed process-handle flavors share the common concept to perform async polling for process-termination and might be the base for a generalized process-handle concept in Rust `std::process`, leveraging a strong ownership-concept for child-processes. The handling of signal SIGCHLD and sharing of child-process meta-data would no longer be required.
```
